discovery-azure-api/src/main/scala/akka/discovery/azureapi/rbac/aks/AzureRbacAksServiceDiscovery.scala

```diff
@@ -0,0 +1,258 @@
+      itemSpec <- item.spec.toSeq
+      itemStatus <- item.status.toSeq
+      if itemStatus.phase.contains("Running")
+      if containerName.forall(name =>
+        itemStatus.containerStatuses match {
+          case Some(statuses) => statuses.filter(_.name == name).exists(!_.state.contains("waiting"))
+          case None           => false
+        })
+      ip <- itemStatus.podIP.toSeq
+      // Maybe port is an Option of a port, and will be None if no portName was requested
+      maybePort <- portName match {
+        case None =>
+          List(None)
+        case Some(name) =>
+          for {
+            container <- itemSpec.containers
+            ports <- container.ports.toSeq
+            port <- ports
+            if port.name.contains(name)
+          } yield Some(port.containerPort)
+      }
+    } yield {
+      val hostOrIp = if (rawIp) ip else s"${ip.replace('.', '-')}.$podNamespace.pod.$podDomain"
+      ResolvedTarget(
+        host = hostOrIp,
+        port = maybePort,
+        address = Some(InetAddress.getByName(ip))
+      )
+    }
+
+  private final class KubernetesApiException(msg: String) extends RuntimeException(msg) with NoStackTrace
+
+  private final class AzureIdentityException(msg: String) extends RuntimeException(msg) with NoStackTrace
+
+  private final case class KubernetesSetup(namespace: String, ctx: HttpsConnectionContext)
+}
+
+/**
+ * INTERNAL API
+ *
+ * Finds relevant targets given a pod list. Note that this doesn't filter by name as it is the job of the selector
+ * to do that.
+ */
+@InternalApi
+final class AzureRbacAksServiceDiscovery(implicit system: ExtendedActorSystem) extends ServiceDiscovery {
+  private val http = Http()
+
+  private val settings = Settings(system.settings.config.getConfig("akka.discovery.azure-rbac-aks-api"))
+
+  private val log = Logging(system, classOf[AzureRbacAksServiceDiscovery])
+
+  private val kubernetesSetup = {
+    implicit val blockingDispatcher: ExecutionContext = system.dispatchers.lookup(DefaultBlockingDispatcherId)
+    for {
+      namespace: String <- Future {
+        settings.podNamespace
+          .orElse(readConfigVarFromFilesystem(settings.podNamespacePath, "pod-namespace"))
+          .getOrElse("default")
+      }
+      httpsContext <- Future(clientHttpsConnectionContext())
+    } yield {
+      KubernetesSetup(namespace, httpsContext)
+    }
+  }
+
+  log.debug("Settings {}", settings)
+
+  import system.dispatcher
+
+  private def fetchAccessToken: Future[AccessToken] =
+    azureDefaultCredential
+      .getToken(accessTokenRequestContext.addScopes(settings.entraServerId))
+      .onErrorMap { error =>
+        log.error("[{}]", error)
+        new AzureIdentityException(
+          """
+          |Attempt failed while fetching access token. Check if workload identity is enabled for the cluster or not and
+          |if the pods has been injected with required AZURE environment variables
+          |"""".stripMargin)
```

Comment:
Don't use line breaks in exceptions or log messages, because line breaks are often handled as delimiters in logs.

Reply:
Got it.